crates/sozo/ops/src/migrate/mod.rs

```diff
@@ -458,63 +522,79 @@ where
+            ui.update_text_boxed(ui_text);
 
-            for (idx, (_, labeled_class)) in classes.into_iter().enumerate() {
-                let declarer_idx = idx % declarers.len();
-                declarers[declarer_idx].add_class(labeled_class);
-            }
+            invoker.invoke_all_sequentially().await?;
+        }
 
-            let ui_text =
-                format!("Declaring {} classes with {} accounts...", n_classes, declarers.len());
-            ui.update_text_boxed(ui_text);
+        Ok(has_changed)
+    }
 
-            let declarers_futures =
-                futures::future::join_all(declarers.into_iter().map(|d| d.declare_all())).await;
+    /// Syncs the external contracts by declaring their classes and deploying them with
+    /// configured constructor data.
+    ///
+    /// Returns true if at least one external contract has changed, false otherwise.
+    async fn sync_external_contracts(
+        &self,
+        ui: &mut MigrationUi,
+    ) -> Result<bool, MigrationError<A::SignError>> {
+        let ui_text =
+            format!("Syncing {} external contracts...", self.diff.external_contracts.len());
+        ui.update_text_boxed(ui_text);
 
-            for declarer_results in declarers_futures {
-                if let Err(e) = declarer_results {
-                    // The issue is that `e` is bound to concrete type `SingleOwnerAccount`.
-                    // Thus, we can't return `e` directly.
-                    // Might have a better solution by addind a new variant?
-                    if e.to_string().contains("Class already declared") {
-                        // If the class is already declared, it might be because it was already
-                        // declared in a previous run or an other declarer.
-                        continue;
-                    }
+        let mut invoker = Invoker::new(&self.world.account, self.txn_config);
 
-                    return Err(MigrationError::DeclareClassError(e.to_string()));
+        // declaring external contract classes
+        let classes: HashMap<_, _> = self
+            .diff
+            .external_contract_classes
+            .iter()
+            .filter_map(|(_, c)| self.external_contract_classes(c))
+            .collect();
+
+        let ui_text = format!("Declaring {} external contract classes...", classes.len());
+        ui.update_text_boxed(ui_text);
+
+        self.declare_classes(ui, classes).await?;
+
+        // then deploying new external contracts
+        let deployer = Deployer::new(&self.world.account, self.txn_config);
+
+        for contract in &self.diff.external_contracts {
+            if let ExternalContractDiff::Created(contract) = contract {
+                if let Some((_, call)) = deployer
+                    .deploy_via_udc_getcall(
+                        contract.class_hash,
+                        contract.salt,
+                        &contract.raw_constructor_data,
```

Comment:
If two contracts have different instance name, but same contract name, same salt and same constructor arguments, it will fail because the address doesn't take in account the instance name.

You could hash the instance name and the salt to have the final salt. This ensures unicity based on the instance name, which is what instance name is for from what I understand. 👍

Reply:
Yes, good idea! Thanks 🙏